docs/HowToGuides/GettingStarted.md

```diff
@@ -0,0 +1,593 @@
+
+1. Create a directory for the whole project:
+   ```sh
+   mkdir -p swift-project/swift
+   cd swift-project/swift
+   ```
+2. Clone the sources:
+   - Via SSH (recommended):
+     If you plan on regularly making direct commits,
+     cloning over SSH provides a better experience.
+     After you've [uploaded your SSH keys to GitHub][]:
+     ```sh
+     git clone git@github.com:apple/swift.git .
+     utils/update-checkout --clone-with-ssh
+     ```
+   - Via HTTPS:
+     If you want to check out the sources as read-only,
+     or are not familiar with setting up SSH,
+     you can use HTTPS instead:
+     ```sh
+     git clone https://github.com/apple/swift.git .
+     utils/update-checkout --clone
+     ```
+   **Note:** If you've already forked the project on GitHub at this stage,
+   **do not clone your fork** to start off. We describe
+   [how to setup your fork](#setting-up-your-fork) in a subsection below.
+3. Double-check that `swift`'s sibling directories are present.
+   ```sh
+   ls ..
+   ```
+   This should list directories like `llvm-project`, `swiftpm` and so on.
+4. Checkout the right branch/tag:
+   If you are building the toolchain for local development, you can skip this
+   step, as Step 2 will checkout `swift`'s `master` branch and matching
+   branches for other projects. <!-- [TODO: master-to-main-rename] -->
+   If you are building the toolchain as a one-off, it is more likely that you
+   want a specific branch or a tag, often corresponding to a specific release
+   or a specific snapshot. You can update the branch/tag for all repositories
+   as follows:
+   ```sh
+   utils/update-checkout --branch mybranchname
+   # OR
+   utils/update-checkout --tag mytagname
+   ```
+   Detailed branching information, including names for release branches, can
+   be found in [Branches.md](/docs/Branches.md).
+
+**Note:**
+The commands used in the rest of this guide assumes that the absolute path
+to your working directory is something like `/path/to/swift-project/swift`.
+Double-check that running `pwd` prints a path ending with `swift`.
+
+[uploaded your SSH keys to GitHub]: https://help.github.com/articles/adding-a-new-ssh-key-to-your-github-account/
+
+### Setting up your fork
+
+If you are building the toolchain for development and submitting patches,
+you will need to setup a GitHub fork.
+
+First fork the `apple/swift` [repository](https://github.com/apple/swift.git),
+using the "Fork" button in the web UI, near the top-right. This will create a
+repository `username/swift` for your GitHub username. Next, add it as a remote:
+```sh
+# Using 'my-remote' as a placeholder name.
+
+# If you set up SSH in step 2
+git remote add my-remote git@github.com:username/swift.git
+
+# If you used HTTPS in step 2
+git remote add my-remote https://github.com/username/swift.git
+```
+Finally, create a new branch.
+```sh
+# Using 'my-branch' as a placeholder name
+git checkout my-branch
+git push --set-upstream my-remote my-branch
+```
```

Comment:
First, sorry for marking such a large block of text!

I find myself confused while reading the _Cloning the project_ section. Partially because it's different from my experience, where I cloned my fork directly with no problems; and partially because it seems contradictory to first-time contributors' lack of commit access that prevent them from making direct commits.

Would the following changes improve the instructions by targeting first-time contributors better:

1. Mention that a repository can be cloned via either SSH or HTTPS, and what the pros and cons are for each.
2. Provide instructions for setting up a fork first, before those for cloning `apple/swift` directly.
3. Remove the warning for against cloning the fork. At least, [according to GitHub](https://guides.github.com/activities/forking/), this should be ok.
4. Add a step of setting up the upstream remote 
  `git remote add my-upstream-remote https://github.com/apple/swift.git`.
5. Follow the instructions for setting up a fork, with those for cloning `apple/swift` directly.

---

^ After writing the list above, I realised that the current instructions are basically the same as my proposed changes. The only difference is that the current instructions set up a remote for `apple/swift` first before the fork. But just in case the description of my initial confusion add some value as a feedback, I think I should leave it up here.

Reply:
> Partially because it's different from my experience, where I cloned my fork directly with no problems

This is based on a [recent issue](https://forums.swift.org/t/problems-with-building-the-swift-compiler/39833/4) someone faced. While it may have worked for you, it didn't work for someone, so I think it's better to be on the safe side here until we've figured out/fixed the underlying issue.

I will add an HTML comment here indicating why this is the way it is.

> partially because it seems contradictory to first-time contributors' lack of commit access that prevent them from making direct commits

"direct commits" is incorrect phrasing, I will change that. The intention is "contributing regularly".

Reply:
I just tried `util/update-checkout`, and it didn't work. I was wrong.

Reply:
Could you add information about hashes (`utils/update-checkout --dump-hashes`) and remotes to https://bugs.swift.org/browse/SR-13476 and provide more information on how to reproduce it? That would be great. I tried reproducing it with a fork but couldn't.

Reply:
I didn't get exactly the same error as the one on the forums. Should I create a separate bug report for it instead?

Reply:
Sure, opening a new bug report also works. We can link the two later if needed.

Reply:
Just filed [SR-13505](https://bugs.swift.org/browse/SR-13505).